File: 0518-coin-change-ii/0518-coin-change-ii.java
class Solution {
    public int change(int amount, int[] coins) {
        int[][] dp = new int[coins.length+1][amount+1];
        for(int i=0; i<dp.length; ++i) dp[i][0] = 1;
        for(int l=1; l<dp.length; ++l){
            for(int t=1; t<dp[0].length; ++t){
                if(t>=coins[l-1]) dp[l][t] = dp[l-1][t] + dp[l][t-coins[l-1]];
                else dp[l][t] = dp[l-1][t];
            }
        }
        return dp[coins.length][amount];
    }
}


// class Solution {
//     public int change(int amount, int[] coins) {
//         int[][] dp = new int[coins.length+1][amount+1];
//         for(int l=0; l<dp.length; ++l){
//             for(int t=0; t<dp[0].length; ++t){
//                 if(t==0) dp[l][t] = 1;
//                 else if(l==0) dp[l][t] = 0;
//                 else{
//                     if(t>=coins[l-1]) dp[l][t] = dp[l-1][t] + dp[l][t-coins[l-1]];
//                     else dp[l][t] = dp[l-1][t];
//                 }
//             }
//         }
//         return dp[coins.length][amount];
//     }
// }

File: 0518-coin-change-ii/0518-coin-change-ii.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        dp = [[0]*(len(coins)+1) for _ in range(amount+1)]
        for t in range(amount+1):
            for l in range(len(coins)+1):
                if t==0: dp[t][l] = 1
                elif l==0: dp[t][l] = 0
                else:
                    dp[t][l] = dp[t][l-1]
                    if t>=coins[l-1]: dp[t][l] += dp[t-coins[l-1]][l]
        return dp[-1][-1]



# class Solution:
#     def change(self, amount: int, coins: List[int]) -> int:
#         dp = [[0]*(len(coins)+1) for _ in range(amount+1)]
#         for t in range(amount+1):
#             for l in range(len(coins)+1):
#                 if t==0: dp[t][l] = 1
#                 elif l==0: dp[t][l] = 0
#                 else:
#                     if t>=coins[l-1]: dp[t][l] = dp[t][l-1] + dp[t-coins[l-1]][l]
#                     else: dp[t][l] = dp[t][l-1]
#         return dp[-1][-1]